src/qt/nametablemodel.cpp

```diff
@@ -0,0 +1,527 @@
+                *outNewRowIndex = parent->index(lowerIndex, 0).row();
+            break;
+        case CT_UPDATED:
+            if (!inModel)
+            {
+                LogPrintf ("Warning: NameTablePriv::updateEntry: Got CT_UPDATED, but entry is not in model\n");
+                break;
+            }
+            lower->name = name;
+            lower->value = value;
+            lower->nHeight = nHeight;
+            parent->emitDataChanged(lowerIndex);
+            break;
+        case CT_DELETED:
+            if (!inModel)
+            {
+                LogPrintf ("Warning: NameTablePriv::updateEntry: Got CT_DELETED, but entry is not in model\n");
+                break;
+            }
+            parent->beginRemoveRows(QModelIndex(), lowerIndex, upperIndex-1);
+            cachedNameTable.erase(lower, upper);
+            parent->endRemoveRows();
+            break;
+        }
+    }
+
+    int size()
+    {
+        return cachedNameTable.size();
+    }
+
+    NameTableEntry *index(int idx)
+    {
+        if (idx >= 0 && idx < cachedNameTable.size())
+        {
+            return &cachedNameTable[idx];
+        }
+        else
+        {
+            return NULL;
+        }
+    }
+};
+
+NameTableModel::NameTableModel(const PlatformStyle *platformStyle, CWallet* wallet, WalletModel *parent):
+        QAbstractTableModel(parent),
+        wallet(wallet),
+        walletModel(parent),
+        priv(new NameTablePriv(wallet, this)),
+        platformStyle(platformStyle)
+{
+    columns << tr("Name") << tr("Value") << tr("Expires in");
+    priv = new NameTablePriv(wallet, this);
+    priv->refreshNameTable();
+
+    QTimer *timer = new QTimer(this);
+    connect(timer, SIGNAL(timeout()), this, SLOT(updateExpiration()));
+    timer->start(MODEL_UPDATE_DELAY);
+
+    subscribeToCoreSignals();
+}
+
+NameTableModel::~NameTableModel()
+{
+    unsubscribeFromCoreSignals();
+    delete priv;
+}
+
+void NameTableModel::updateExpiration()
+{
+    int nBestHeight = chainActive.Height();
+    if (nBestHeight != cachedNumBlocks)
+    {
+        LOCK(cs_main);
+
+        cachedNumBlocks = nBestHeight;
+        // Blocks came in since last poll.
+        // Delete expired names
+        for (int i = 0, n = priv->size(); i < n; i++)
+        {
+            NameTableEntry *item = priv->index(i);
+            if (!item->HeightValid())
+                continue;       // Currently, unconfirmed names do not expire in the table
+            int nHeight = item->nHeight;
+
+            // NOTE: the line below used to be: GetExpirationDepth(nHeight)
+            // I changed it to just nHeight for now
+            // int GetExpirationDepth(int nHeight) {
+            //     if (nHeight < 24000)
+            //         return 12000;
+            //     if (nHeight < 48000)
+            //         return nHeight - 12000;
+            //     return 36000;
+            // }
+            if (nHeight + 36000 - nBestHeight <= 0)
```

Comment:
Please use `NameExpirationDepth` similar to the code at https://github.com/domob1812/namecore/blob/master/src/names/main.cpp#L29 (or even use that function directly) instead of hardcoding the 36000 here.